### PR TITLE
feat: support healthcheck field in service config

### PR DESCRIPTION
**Pull request by opjt**
- related issue : https://github.com/containerd/nerdctl/issues/4709

 ## Summary

`services.<SERVICE>.healthcheck` was listed as unimplemented — this PR wires it up.

compose-go already parses the `healthcheck` block, so this is mostly connecting
it to the existing `--health-*` flags in serviceparser.

**Limitations:**
- `CMD` exec form is converted to `CMD-SHELL` since `--health-cmd` always stores the command as `CMD-SHELL` (see `withHealthcheck` in `pkg/cmd/container/create.go`)
- `start_interval` is not supported — I plan to work on this after this PR is merged.

 ## Test plan

  ```yaml
  services:
    web:
      image: nginx:alpine
      healthcheck:
        test: ["CMD-SHELL", "curl -f http://localhost"]
        interval: 10s
        timeout: 5s
        retries: 3
        start_period: 2s
```
```bash
  nerdctl compose up -d
  nerdctl inspect <container-name> | grep health
```

### CODE REVIEW

**Comment by opjt** [comment]
Hi, just following up on this PR. Thanks!

File: pkg/composer/serviceparser/serviceparser.go [hunk 1]
```diff
@@ -747,6 +764,44 @@ func newContainer(project *types.Project, parsed *Service, i int) (*Container, e
 		c.RunArgs = append(c.RunArgs, "-w="+svc.WorkingDir)
 	}
 
+	if svc.HealthCheck != nil {
+		hc := svc.HealthCheck
+		if hc.Disable {
+			c.RunArgs = append(c.RunArgs, "--no-healthcheck")
+		} else {
+			if len(hc.Test) > 0 {
+				switch hc.Test[0] {
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser.go]: The `default` statement is missing, so please add it and modify the code to output a warning.

File: pkg/composer/serviceparser/serviceparser.go [hunk 2]
```diff
@@ -747,6 +764,44 @@ func newContainer(project *types.Project, parsed *Service, i int) (*Container, e
...
+					// CMD exec form is converted to CMD-SHELL because --health-cmd always stores
+					// the command as CMD-SHELL (see pkg/cmd/container/create.go: withHealthcheck).
+					// This means the command will be executed via /bin/sh -c instead of exec directly.
+					if len(hc.Test) >= 2 {
+						log.L.Warnf("service %s: healthcheck: CMD exec form is not supported, converting to CMD-SHELL", svc.Name)
+						c.RunArgs = append(c.RunArgs, fmt.Sprintf("--health-cmd=%s", strings.Join(hc.Test[1:], " ")))
+					}
+				}
+			}
+			if hc.Interval != nil {
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser.go:792]: Although flags such as `--health-interval` are added to the command when specifying `test: ["NONE"]`, they should not be included.

File: pkg/composer/serviceparser/serviceparser.go [hunk 3]
```diff
@@ -747,6 +764,44 @@ func newContainer(project *types.Project, parsed *Service, i int) (*Container, e
...
+		hc := svc.HealthCheck
+		if hc.Disable {
+			c.RunArgs = append(c.RunArgs, "--no-healthcheck")
+		} else {
+			if len(hc.Test) > 0 {
+				switch hc.Test[0] {
+				case healthcheck.CmdNone:
+					c.RunArgs = append(c.RunArgs, "--no-healthcheck")
+				case healthcheck.CmdShell:
+					if len(hc.Test) >= 2 {
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser.go]: Warning messages should be added `if len(hc.Test) < 2`. WDYT?.

File: pkg/composer/serviceparser/serviceparser.go [hunk 4]
```diff
@@ -747,6 +764,44 @@ func newContainer(project *types.Project, parsed *Service, i int) (*Container, e
...
+						c.RunArgs = append(c.RunArgs, fmt.Sprintf("--health-cmd=%s", hc.Test[1]))
+					}
+				case healthcheck.Cmd:
+					// CMD exec form is converted to CMD-SHELL because --health-cmd always stores
+					// the command as CMD-SHELL (see pkg/cmd/container/create.go: withHealthcheck).
+					// This means the command will be executed via /bin/sh -c instead of exec directly.
+					if len(hc.Test) >= 2 {
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser.go]: Warning messages should be added `if len(hc.Test) < 2`. WDYT?.

File: cmd/nerdctl/compose/compose_up_linux_test.go [hunk 5]
```diff
@@ -1314,3 +1315,61 @@ services:
 
 	testCase.Run(t)
 }
+
+func TestComposeUpHealthcheck(t *testing.T) {
```
Comment (haytok) [inline cmd/nerdctl/compose/compose_up_linux_test.go:1319]: IMO, tests in case of `test: ["NONE"]` and `disable: true` should be added.

- https://github.com/compose-spec/compose-spec/blob/main/spec.md#healthcheck

File: pkg/composer/serviceparser/serviceparser_test.go [hunk 6]
```diff
@@ -623,3 +623,64 @@ services:
...
+	project, err := testutil.LoadProject(comp.YAMLFullPath(), comp.ProjectName(), nil)
+	assert.NilError(t, err)
+
+	getContainersFromService := func(svcName string) []Container {
+		svcConfig, err := project.GetService(svcName)
+		assert.NilError(t, err)
+		svc, err := Parse(project, svcConfig)
+		assert.NilError(t, err)
+		return svc.Containers
+	}
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser_test.go]: This function can be used in `TestParseRestartPolicy`, so that should be defined in outer scope.

File: pkg/composer/serviceparser/serviceparser.go (hunk 1 above)
Comment (opjt) [reply pkg/composer/serviceparser/serviceparser.go]: I'll add the `default` case with a warning log.

File: pkg/composer/serviceparser/serviceparser_test.go (hunk 6 above)
Comment (opjt) [reply pkg/composer/serviceparser/serviceparser_test.go]: I'll move it to the outer scope. thanks

File: cmd/nerdctl/compose/compose_up_linux_test.go (hunk 5 above)
Comment (opjt) [reply cmd/nerdctl/compose/compose_up_linux_test.go:1319]: Since `serviceparser_test` already covers both `test: ["NONE"]` and `disable: true` mapping to `--no-healthcheck`, I think adding just one of them in the test is sufficient. WDYT?

File: pkg/composer/serviceparser/serviceparser.go (hunk 2 above)
Comment (opjt) [reply pkg/composer/serviceparser/serviceparser.go:792]: I will fix it to avoid appending unnecessary flags.

File: pkg/composer/serviceparser/serviceparser.go (hunk 1 above)
Comment (opjt) [reply pkg/composer/serviceparser/serviceparser.go]: The `default` case is unreachable since `compose-go` already validates the healthcheck test type.
I'll remove it

<https://github.com/compose-spec/compose-go/blob/v2.10.2/loader/validate.go#L80>

File: pkg/composer/serviceparser/serviceparser.go (hunk 3 above)
Comment (opjt) [reply pkg/composer/serviceparser/serviceparser.go]: Do we really need to warn the user about what is simply a user error?
Docker silently ignores this case, so I'm not sure which direction we should go in this project.

Comment (haytok) [reply pkg/composer/serviceparser/serviceparser.go]: Since nothing was displayed in Docker when `len(hc.Test) == 1`, it’s best to leave it as is in nerdctl as well. Thanks!

File: pkg/composer/serviceparser/serviceparser_test.go [hunk 7]
```diff
@@ -601,25 +610,69 @@ services:
...
+	assert.Assert(t, in(c.RunArgs, "--health-retries=3"))
+	assert.Assert(t, in(c.RunArgs, "--health-start-period=5s"))
+
+	c = getContainersFromService(t, project, "cmd_exec")[0]
+	assert.Assert(t, in(c.RunArgs, "--health-cmd=curl -f http://localhost"))
+	assert.Assert(t, in(c.RunArgs, "--health-interval=1m0s"))
+
+	c = getContainersFromService(t, project, "disabled_flag")[0]
+	assert.Assert(t, in(c.RunArgs, "--no-healthcheck"))
+	assert.Assert(t, !in(c.RunArgs, "--health-cmd="))
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser_test.go]: `in` is `strutil.InStringSlice` (exact element match), so `in(c.RunArgs, "--health-cmd=")` never matches the actual element (`--health-cmd=<value>`). Shouldn't we check whether any of the argument elements have a prefix of `--health-cmd=`?

File: pkg/composer/serviceparser/serviceparser.go [hunk 8]
```diff
@@ -747,6 +764,46 @@ func newContainer(project *types.Project, parsed *Service, i int) (*Container, e
 		c.RunArgs = append(c.RunArgs, "-w="+svc.WorkingDir)
 	}
 
+	if svc.HealthCheck != nil {
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser.go:767]: When both `disable: true` and `test: [...]` are specified like below, the switch still appends `--health-cmd=...` before the if disabled check, so nerdctl run ends up with both `--health-cmd=...` and `--no-healthcheck`.

```yaml
services:
  web:
    image: nginx:alpine
    healthcheck:
      test: ["CMD-SHELL", "curl -f http://localhost"]
      interval: 1m
      timeout: 5s
      retries: 3
      start_period: 2s
      disable: true
```

In this case, `test` should be ignored.

File: pkg/composer/serviceparser/serviceparser.go [hunk 9]
```diff
@@ -121,6 +123,21 @@ func warnUnknownFields(svc types.ServiceConfig) {
...
+		if unknown := reflectutil.UnknownNonEmptyFields(svc.HealthCheck,
+			"Test",
+			"Timeout",
+			"Interval",
+			"Retries",
+			"StartPeriod",
+			"Disable",
+			"Extensions",
+			// TODO: add support 'startinterval'
```
Comment (haytok) [inline pkg/composer/serviceparser/serviceparser.go]: NIT:
```suggestion
			// TODO: add support 'StartInterval'
```

File: pkg/composer/serviceparser/serviceparser.go (hunk 8 above)
Comment (opjt) [reply pkg/composer/serviceparser/serviceparser.go:767]: I'll fix this and update the test case

**Review by haytok — APPROVED** [review]
LGTM, Thanks!

**Review by AkihiroSuda — APPROVED** [review]
Thanks
